### ISSUE TITLE
404 when trying to accept notification

### ISSUE THREAD

**Issue by fafische** [nextcloud/twofactor_nextcloud_notification#773]
### Steps to reproduce
1. Login with username and password
2. select "Nextcloud-Benachrichtigung" as two-factor mode
3. click to "Bestätigen" in notification in nextcloud app

### Expected behaviour
user login happens, notification disappears

### Actual behaviour
No login, notifications stays

### Server configuration
**Operating system**:
Archlinux

**Web server:** Apache

**Database:** 
Art: mysql
Version: 11.1.2

**PHP version:**
8.1.24

**Nextcloud version:** (see Nextcloud admin page)
[Nextcloud Hub 6] (https://nextcloud.com/) (27.1.3)

**Where did you install Nextcloud from:**
manual installation, integrated updater

**List of activated apps:**

```
Enabled:
  - activity: 2.19.0
  - calendar: 4.5.3
  - carnet: 0.25.2
  - circles: 27.0.1
  - cloud_federation_api: 1.10.0
  - comments: 1.17.0
  - contacts: 5.4.2
  - contactsinteraction: 1.8.0
  - dashboard: 7.7.0
  - dav: 1.27.0
  - federatedfilesharing: 1.17.0
  - federation: 1.17.0
  - files: 1.22.0
  - files_automatedtagging: 1.17.0
  - files_external: 1.19.0
  - files_pdfviewer: 2.8.0
  - files_reminders: 1.0.0
  - files_rightclick: 1.6.0
  - files_sharing: 1.19.0
  - files_trashbin: 1.17.0
  - files_versions: 1.20.0
  - firstrunwizard: 2.16.0
  - keeweb: 0.6.13
  - logreader: 2.12.0
  - lookup_server_connector: 1.15.0
  - nextcloud_announcements: 1.16.0
  - notifications: 2.15.0
  - oauth2: 1.15.1
  - password_policy: 1.17.0
  - photos: 2.3.0
  - previewgenerator: 5.3.0
  - privacy: 1.11.0
  - provisioning_api: 1.17.0
  - recommendations: 1.6.0
  - related_resources: 1.2.0
  - richdocuments: 8.2.2
  - richdocumentscode: 23.5.503
  - serverinfo: 1.17.0
  - settings: 1.9.0
  - sharebymail: 1.17.0
  - spreed: 17.1.2
  - support: 1.10.0
  - survey_client: 1.15.0
  - systemtags: 1.17.0
  - tasks: 0.15.0
  - text: 3.8.0
  - theming: 2.2.0
  - twofactor_backupcodes: 1.16.0
  - twofactor_nextcloud_notification: 3.8.0
  - twofactor_totp: 9.0.0
  - updatenotification: 1.17.0
  - user_status: 1.7.0
  - viewer: 2.1.0
  - weather_status: 1.7.0
  - workflowengine: 2.9.0
Disabled:
  - admin_audit: 1.17.0
  - bruteforcesettings: 2.7.0 (installed 2.4.0)
  - cospend: 1.5.10 (installed 1.5.10)
  - encryption: 2.15.0
  - files_fulltextsearch_tesseract: 27.0.0 (installed 27.0.0)
  - suspicious_login: 5.0.0
  - user_ldap: 1.17.0
```

**Nextcloud configuration:**

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "27.1.3.2",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "forcessl": true,
        "logtimezone": "Europe\/Berlin",
        "log_rotate_size": 10485760,
        "preview_libreoffice_path": "\/usr\/bin\/libreoffice",
        "preview_office_cl_parameters": "",
        "preview_max_x": "2048",
        "preview_max_y": "2048",
        "enable_previews": true,
        "enabledPreviewProviders": [
            "OC\\Preview\\SGI",
            "OC\\Preview\\HEIC",
            "OC\\Preview\\BMP",
            "OC\\Preview\\GIF",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\PNG",
            "OC\\Preview\\TXT",
            "OC\\Preview\\XBitmap"
        ],
        "theme": "",
        "maintenance": false,
        "trusted_domains": [
            "nextcloud.***REMOVED SENSITIVE VALUE***.org",
            "homeserver.***REMOVED SENSITIVE VALUE***.org",
            "192.168.5.12"
        ],
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "loglevel": 2,
        "overwrite.cli.url": "https:\/\/***REMOVED SENSITIVE VALUE***.org\/owncloud",
        "htaccess.RewriteBase": "\/owncloud",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trashbin_retention_obligation": "auto",
        "updater.release.channel": "stable",
        "mysql.utf8mb4": true,
        "app_install_overwrite": [
            "calendar",
            "keeweb",
            "carnet"
        ],
        "has_rebuilt_cache": true,
        "jpeg_quality": "90",
        "enable_movie_transcode": false,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "dbindex": 0,
            "password": "***REMOVED SENSITIVE VALUE***",
            "timeout": 1.5
        },
        "default_phone_region": "DE",
        "mail_sendmailmode": "pipe"
    }
}

```

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
No

### Client configuration
**Browser:**
Nexcloud 3.26.0 on Android
and
Nexcloud Desktop CLient 3.10.0 on Windows

#### Nextcloud log (data/nextcloud.log)
no entrys at time

#### Apache
[10/Nov/2023:12:21:15 +0100] "POST /owncloud/ocs/v2.phpud/ocsapp/apps/twofactor_nextcloud_notification/api/v1/attempt/408 HTTP/1.1" 404 13384

**Comment by fafische** [comment]
Looks like there is an error in the link to submit result when clicking in notification:
ocs/v2.phpud/.. instead of ocs/v2.php/..

**Comment by skug67** [comment]
I'm not seeing the same error message, but since updating to 3.8.0, the app is no longer working.  When I try to approve a login from my iPad nothing happens, and when I try on my Android phone I get a message that execution failed.

**Comment by tonyskilton** [comment]
I'm seeing similar since updating
The notification comes through to my phone, I confirm, and the notification repeats/reappears. without authenticating the login.
This just loops until I cancel the login.

**Comment by nickvergessen** [comment]
I will have a look later today, but it worked in the test I did before the release. Seeing:
```
        "overwrite.cli.url": "https:\/\/***REMOVED SENSITIVE VALUE***.org\/owncloud",
        "htaccess.RewriteBase": "\/owncloud",
```

There might be something wrong with paths. I will try to get that locally, but the app uses public Nextcloud API so the problem should exist in a lot of locations.

**Comment by skug67** [comment]
I've been seeing other errors in my apache logs (none of which appear to impact usability) that do indicate those rewrite functions are somehow being ignored at times.

**Comment by nickvergessen** [comment]
I'm unable to reproduce this locally.

![Bildschirmfoto vom 2023-11-13 12-10-43](https://github.com/nextcloud/twofactor_nextcloud_notification/assets/213943/dbe33bea-b57b-4576-b940-184a761b6948)

But it's a bit confusing that only this app here would break. The method should be used by any notification provider, activity provider and more.

Can you check if `occ maintenance:update:htaccess` fixes the issue? (Ref https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/occ_command.html)

**Comment by skug67** [comment]
Have run the htaccess update script already.  I do suspect that the problem is more broadly with that redirection.  I may try to see if I can spin up a new subdomain and run at the root of that to see if it fixes the problem.  But that may take a few days (since I've got a real job to do in the meantime.....)

**Comment by skug67** [comment]
I actually just upped the loglevel for apache and discovered that when I clicked the notification it was trying to reach the following (non-existent) URL:

ocs/v2.phpoud2/ocsapp/apps/twofactor_nextcloud_notification/api/v1/attempt/308

I do note that I run my installation at the suburl "/nextcloud2" so my "oud2" is adding the final 4 characters of that URL, while @fafische was getting an added "ud" (the final two characters) of "nextcloud"  It leads me to believe that there is something pretty funky going on the with the Rewrite function.  But I sure don't know what it is.....

**Comment by skug67** [comment]
Actually reviewing the apache logs more carefully, it looks to me like (subject to the important caveat that I'm way out my depth here) it's the app itself and not the .htaccess Rewrite rules that is adding the extra characters into the link.

**Comment by nickvergessen** [comment]
> it looks to me like (subject to the important caveat that I'm way out my depth here) it's the app itself

I'm in my depth and can tell you that `IURLGenerator::linkToOCSRouteAbsolute()` is exactly what should be used.
https://github.com/nextcloud/twofactor_nextcloud_notification/blob/main/lib/Notification/Notifier.php#L81-L82

> my installation at the suburl "/nextcloud2" so my "oud2" is adding the final 4 characters of that URL, 

On the initial report it's the last 3 only, but actually it's "The full subpath without the first 6 characters"

Could one of you create a test account on their instance and give me access? You could use `<my github name>@nextcloud.com` to reach out with the personal URLs

**Comment by zorgluf** [comment]
Account created, did you received the email ?
In my case, the wrong url is : POST /ocs/v2.phpapp/apps/twofactor_nextcloud_notification/api/v1/attempt/92
The directory rewrite base is /nc.
Note that I am using the nextcloud official docker image (apache base).

**Comment by nonplusnl** [comment]
Same here, logging in using notification is failing since new release. See Apache log below.
First a log entry from the device that wants to login, then a log entry from phone, confirming notification via Android app. You see the 404 generated, but also the different URL: .../ocs/v2.phpsapp/ vs /npc/ocs/v2.php/

IPv6-ADDR_DEVICE_wanting_to_log_in - - [18/Nov/2023:20:22:25 +0100] "GET /SUBDOMAIN/ocs/v2.php/apps/twofactor_nextcloud_notification/api/v1/poll/o6OEDZoYdYiMy6AuOzoztUOHmKXfscWXcUj8MlMB HTTP/1.1" 200 974
IPv6-ADDR_DEVICE_used_to_confirm_notification - USERNAME [18/Nov/2023:20:22:26 +0100] "POST /SUBDOMAIN/ocs/v2.phpsapp/apps/twofactor_nextcloud_notification/api/v1/attempt/318 HTTP/1.1" 404 19774

**Comment by nickvergessen** [comment]
So with some looking around, I could reproduce it locally after enabling this system config:
```php
'htaccess.IgnoreFrontController' => true,
```

But as per my initial comment this does break all OCS routes, so it is a bug in server.

**Comment by nickvergessen** [comment]
Fix is in https://github.com/nextcloud/server/pull/41616

See https://docs.nextcloud.com/server/latest/admin_manual/issues/applying_patch.html#getting-a-patch-from-a-github-pull-request for how to apply a patch

**Comment by skug67** [comment]
I had to delete the tests section of the patch (those files are not present on my installation), but the patch to lib/private/URLGenerator.php worked perfecly.  Many thanks.Sent from my iPadOn Nov 20, 2023, at 3:21 AM, Joas Schilling ***@***.***> wrote:﻿
Fix is in nextcloud/server#41616
See https://docs.nextcloud.com/server/latest/admin_manual/issues/applying_patch.html#getting-a-patch-from-a-github-pull-request for how to apply a patch
